analysis/src/TypeUtils.ml

```diff
@@ -109,6 +226,45 @@ let rec extractFunctionType ~env ~package typ =
   in
   loop ~env [] typ
 
+let maybeSetTypeArgCtx ?typeArgContextFromTypeManifest ~typeParams ~typeArgs env
+    =
+  match typeArgContextFromTypeManifest with
+  | Some typeArgContextFromTypeManifest -> Some typeArgContextFromTypeManifest
+  | None ->
+    let typeArgContext =
+      if List.length typeParams > 0 then Some {env; typeParams; typeArgs}
+      else None
+    in
+    (match typeArgContext with
+    | None -> ()
+    | Some typeArgContext ->
+      if Debug.verbose () then
+        Printf.printf "[#type_arg_ctx]--> setting new type arg ctx: %s"
+          (debugLogTypeArgContext typeArgContext));
+    typeArgContext
+
+let rec extractFunctionType2 ?typeArgContext ~env ~package typ =
```

Comment:
Same as existing one but doesn't instantiate. Could potentially remove this one all together in the larger refactor as it's only special because it can do lookup of types.